The apoc.custom.declareProcedure throws exception "Query results do not match requested output" with some queries

## How to Reproduce the Problem

See https://community.neo4j.com/t/apoc-custom-declareprocedure-always-throws-exception-corresponding-cypher-code-works/52004


### Versions

  - OS: Mac OSX
  - Neo4j: 4.3
  - Neo4j-Apoc: 4.3.0.5

Here's an even simpler example of two variants of basically the same procedure declaration, of which one works and the other one doesn't:

**Variant 0:**
```
 CALL apoc.custom.declareProcedure(
    "test() :: (abc::INT)",
    "
    call {
        with 1 as abc return abc
        union all
        with 2 as abc return abc
    }
    return abc
    "
);
```
**RESULT:** Failure - `Failed to invoke procedure apoc.custom.declareProcedure: Caused by: java.lang.RuntimeException: Query results do not match requested output.`

**Variant 1:**
```
 CALL apoc.custom.declareProcedure(
    "test() :: (abc::INT)",
    "
    call {
        with 1 as a return a
        union all
        with 2 as a return a
    }
    return a as abc
    "
);
```
**RESULT:** Success

**Tested with:**
Neo4j: 4.4.7
APOC: 4.4.0.4, 4.4.0.7

**Note** that both of the examples above work normally with `APOC: 4.4.0.3`

We are not able to upgrade from APOC 4.4.0.3 due to the reported issue. Creation of some of our custom procedures fails . Steps to reproduce were described by @ralphavalon above: 

> Here's how to reproduce it:
```
version: '3'
services:
  neo4j:
    image: neo4j:4.4.7
    hostname: neo4j
    ports:
      - "7474:7474"
      - "7687:7687"
    environment:
      - NEO4J_AUTH=neo4j/test
      - NEO4J_dbms_logs_debug_level=DEBUG
      - NEO4JLABS_PLUGINS=["apoc"]
      - NEO4J_dbms_security_procedures_whitelist=apoc.*
      - NEO4J_dbms_security_procedures_unrestricted=apoc.*
 ```
Then, just login into http://localhost:7474/browser and run:
```
CALL apoc.custom.declareProcedure('exampleTest(exampleId::STRING) ::(value::STRING)',
'MATCH (:ExampleNode)
 OPTIONAL MATCH (o:OtherExampleNode {identifier:$exampleId})
 RETURN o.identifier as value
 UNION ALL
 MATCH (n:ExampleNode)
 OPTIONAL MATCH (o:OtherExampleNode {identifier:$exampleId})
 RETURN o.identifier as value
');
```
The result: 
```
Failed to invoke procedure `apoc.custom.declareProcedure`: Caused by: java.lang.RuntimeException: Query results do not match requested output.
```

Last tested with: 

Neo4j: 4.4.10
APOC: 4.4.0.8

--- a/full/src/main/java/apoc/custom/CypherProcedures.java
+++ b/full/src/main/java/apoc/custom/CypherProcedures.java
@@ -183,7 +183,12 @@ public class CypherProcedures {
                                 (map, value) -> map.put(value.name(), null), HashMap::putAll),
                 result -> {
                     if (!DEFAULT_MAP_OUTPUT.equals(output)) {
-                        checkOutputParams(outputSet, result.columns());
+                        // when there are multiple variables with the same name, e.g within an "UNION ALL" Neo4j adds a suffix "@<number>" to distinguish them, 
+                        //  so to check the correctness of the output parameters we must first remove this suffix from the column names
+                        final Set<String> columns = result.columns().stream()
+                                .map(i -> i.replaceFirst("@[0-9]+", "").trim())
+                                .collect(Collectors.toSet());
+                        checkOutputParams(outputSet, columns);
                     }
                     if (!DEFAULT_INPUTS.equals(input)) {
                         checkInputParams(result);
@@ -210,7 +215,7 @@ public class CypherProcedures {
     }
 
 
-    private void checkOutputParams(Set<String> outputSet, List<String> columns) {
+    private void checkOutputParams(Set<String> outputSet, Set<String> columns) {
         if (!Set.copyOf(columns).equals(outputSet)) {
             throw new RuntimeException(ERROR_MISMATCHED_OUTPUTS);
         }

--- a/full/src/test/java/apoc/custom/CypherProceduresTest.java
+++ b/full/src/test/java/apoc/custom/CypherProceduresTest.java
@@ -666,6 +666,39 @@ public class CypherProceduresTest  {
         // when
         TestUtil.singleResultFirstColumn(db, "return custom.answer()");
     }
+    
+    @Test
+    public void testIssue2605() {
+        db.executeTransactionally("CREATE (n:Test {id: 1})-[:has]->(:Log), (n)-[:has]->(:System)");
+        String query = "MATCH (node:Test)-[:has]->(log:Log) WHERE node.id = $id WITH node \n" +
+                "MATCH (node)-[:has]->(log:System) RETURN log, node";
+        db.executeTransactionally("CALL apoc.custom.declareProcedure('testIssue2605(id :: INTEGER ) :: (log :: NODE, node :: NODE)', $query, 'read')", Map.of("query", query));
+
+        // check query 
+        TestUtil.testCall(db, "call custom.testIssue2605(1)", (row) -> {
+            assertEquals(List.of(Label.label("Test")), ((Node) row.get("node")).getLabels());
+            assertEquals(List.of(Label.label("System")), ((Node) row.get("log")).getLabels());
+        });
+
+        // UNION ALL github issue case
+        db.executeTransactionally("CREATE (n:ExampleNode {id: 1}), (:OtherExampleNode {identifier: '1'})");
+        String query2 = "MATCH (:ExampleNode)\n" +
+                " OPTIONAL MATCH (o:OtherExampleNode {identifier:$exampleId})\n" +
+                " RETURN o.identifier as value\n" +
+                " UNION ALL\n" +
+                " MATCH (n:ExampleNode)\n" +
+                " OPTIONAL MATCH (o:OtherExampleNode {identifier:$exampleId})\n" +
+                " RETURN o.identifier as value";
+        db.executeTransactionally("CALL apoc.custom.declareProcedure('exampleTest(exampleId::STRING) ::(value::STRING)', $query, 'read')", Map.of("query", query2));
+        
+        // check query 
+        final String identifier = "1";
+        TestUtil.testResult(db, "call custom.exampleTest($id)", Map.of("id", identifier), (r) -> {
+            assertEquals(identifier, r.next().get("value"));
+            assertEquals(identifier, r.next().get("value"));
+            assertFalse(r.hasNext());
+        });
+    }
 
     @Test
     public void shouldFailWithMismatchedParameters() {

